WordPress/gutenberg
# Refactored Gallery: Add filter for replacing inner gallery component for alternate layouts

## Description
Adds a filter to allow replacing of the inner Gallery component. This will provide a means for plugins to alter the gallery's layout within the editor without having to completely reimplement the `GalleryEdit` component using the `editor.BlockEdit` filter.

A prime example of this would be a masonry layout using a dedicated React masonry component to lay out the inner blocks.

## Why add this filter?

To a user, a different gallery layout is for all intents and purposes, just a style of gallery. If different gallery layouts can be presented as block styles, it will aid discovery and user experimentation as well as simplify the options in the block inserter. 

The catch here is block styles alone will not help us manage the gallery's layout within the editor. In the above example of a masonry layout, we can get away with adding a plugin to enqueue frontend assets. Unfortunately, we need more control within the editor.

The GalleryEdit component is rather complex and presents a barrier to being able to simply swap it out for an alternate via the `editor.BlockEdit` filter. This new filter narrows the scope of the component being replaced making it much easier to augment the gallery.

## How has this been tested?
Manually.

1. Checkout and build this PR branch
2. Create a quick plugin to leverage this filter and swap out the inner gallery component
3. Create a new post and add a gallery, confirm your new component is used instead.

Something along the lines of:
```js
/**
 * WordPress dependencies
 */
import { registerBlockStyle } from '@wordpress/blocks';
import { addFilter } from '@wordpress/hooks';

/**
 * Internal dependencies
 */
import MasonryGallery from './masonry-gallery.js';

registerBlockStyle( 'core/gallery', {
	name: 'masonry',
	label: 'Masonry',
} );

const maybeApplyMasonryLayout = ( CoreGallery ) => {
	return ( props ) => {
		// Return the original gallery component if masonry style not applied.
		if ( ! props.attributes.className?.includes( 'is-style-masonry' ) ) {
			return <CoreGallery { ...props } />;
		}

		return <MasonryGallery />;
	};
};

// Add a filter to conditionally replace the inner gallery component with one
// catering to alternative layouts based on block style.
addFilter(
	'editor.Gallery',
	'jetpack/masonry-gallery',
	maybeApplyMasonryLayout
);
```

## Screenshots 
![GalleryFilter](https://user-images.githubusercontent.com/60436221/108806438-65c46200-75ed-11eb-9047-9dd269f07654.gif)

## Types of changes
New feature

## Checklist:
